fix(compiler): normalize line endings in ICU expansions

petebacondarwin opened the pull request:
The html parser already normalizes line endings (converting
`\r\n` to `\n`) for most text in templates but it was missing
the expressions of ICU expansions.

In ViewEngine backticked literal strings used to define inline
templates were already normalized by the TypeScript parser.
In Ivy we are parsing the raw text of the source file directly so
the line ending need to be manually normalized.

This change in behaviour was causing the computed i18n message
id for ICUs to change depending upon the current line ending settings
of the environment extracting or merging translations.

Closes #36725


----


Capturing the warning message content here:

When parsing an external template file "TestComp" one or more ICU expressions were found that contain `\\r\\n` line-endings.
These would have been normalized to `\\n` but this would change the computed id for their i18n messages.
The line endings were left unnormalized for backward compatibility on this occasion.
In a future release, these line endings will be normalized, which will be a breaking change.
It is suggested that you try opting into this behavior now to see what effect it will have on your message ids by setting `"i18nNormalizeExternalTemplateLineEndingsInICUs": true` in the `angularCompilerOptions` of `tsconfig.json`
The ICU expressions that need normalizing are:

petebacondarwin commented on packages/compiler/test/ml_parser/html_parser_spec.ts:
```diff
@@ -278,6 +312,36 @@ import {humanizeDom, humanizeDomSourceSpans, humanizeLineColumn} from './ast_spe
+          expect(humanizeDom(parsed)).toEqual([
+            [html.Element, 'div', 0],
+            [html.Text, '\n  ', 1],
+            [html.Expansion, '\n    messages.length', 'plural', 1],
```
@AndrewKushnir - note that while the expansion cases have their whitespace trimmed, the expression does not.

I think this is actually a bug, but we cannot change it without breaking all the message ids. So in order to do so we would need to come up with a migration plan for all the translation files that are already relying upon the "wrong" message id.

ayazhafiz replied:
ExpansionCases should not have their whitespace trimmed in this context, correct?

petebacondarwin replied:
ExpansionCases have always had their whitespace trimmed because of these lines:

https://github.com/angular/angular/blob/master/packages/compiler/src/ml_parser/lexer.ts#L626-L631

petebacondarwin commented:
It turns out that things are really broken as they stand, but that also means that this would be a breaking change for a specific scenario...

On a machine where line endings are set to be `\r\n`, there is a difference between inline and external templates. TypeScript will normalise line endings in backtick string literals to be `\n`. So in ViewEngine the line ending doesn't matter. But in external templates (i.e. `templateUrl`) TypeScript does not get a look in, and we process the contents of the file directly, which means that the line endings are taken into account.

As shown in this PR, the only thing that is affected is the ICU `expression` string - the rest of ICUs and in fact the rest of templates already have the line endings normalised when we parse templates.

This is quite broken, right now, since two people working on different machines with different line ending settings would generate different message ids for the same external templates.

The right thing to do is to fix this but then Windows users with `\r\n` line endings in their external templates will find that their ICU message IDs have changed.

petebacondarwin commented:
I added a fixup commit that should make this change non-breaking - and provides a opt-in switch that we can flip in the future.

petebacondarwin commented on packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -96,8 +96,9 @@ export class DecorationAnalyzer {
         /* i18nUseExternalIds */ true, this.bundle.enableI18nLegacyMessageIdFormat,
-        this.moduleResolver, this.cycleAnalyzer, this.refEmitter, NOOP_DEFAULT_IMPORT_RECORDER,
-        NOOP_DEPENDENCY_TRACKER, this.injectableRegistry, /* annotateForClosureCompiler */ false),
+        /* normalizeExternalTemplateLineEndingsInICUs */ false, this.moduleResolver,
```
In ngcc we are always working with templates that are inline, since they have been previously built by ViewEngine, so really this could be `true|false|undefined` here and we would get the same result.

IgorMinar replied:
Instead of undefined could we specify what the default value is if one is not explicitly provided? Often this is an option that makes the api simpler and less ambiguous but we don't realize that we have that option available. Could this be one of those cases?

petebacondarwin replied:
I am not sure what is being asked for here...

Are you saying that the decision about the default (i.e. whether the switch defaults to `true` or `false` when it is not provided) should be done at the highest level possible?

petebacondarwin replied:
All the `TokenizeOptions` non-required, and this gets passed through to there. I felt that it was better to follow this pattern; and in that case it is better for there to be only one place where the `undefined` gets converted to `true` or `false`. See https://github.com/angular/angular/blob/master/packages/compiler/src/ml_parser/lexer.ts#L64-L108

If we make this option required at the top level then you would have a single required option in `TokenizeOptions`, which doesn't feel right.

IgorMinar replied:
ah.. I missed that the `I18nOptions` is an interface.

AndrewKushnir commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -608,7 +622,13 @@ class _Tokenizer {
+    if (this._escapedString || this._normalizeExternalTemplateLineEndingsInICUs) {
+      this._endToken([this._processCarriageReturns(condition)]);
+    } else {
+      // This expression is in an external template and for backward compatibility we are not
```
Hi @petebacondarwin, I did a first review pass (will complete it tomorrow) and I have a quick question on this section: may be we can extend this a bit and log a message that would explain that we come across a case that would be affected by the breaking change (when we decide to flip the default) and encourage opting-in? Something like:
```
  const processedCondition = this._processCarriageReturns(condition);
  if (processedCondition !== condition) {
    // ... log a message here ...
  }
```

petebacondarwin replied:
Great idea!

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -258,6 +258,11 @@ export class ComponentDecoratorHandler implements
+    // Report any informational warnings to the developer
+    for (const warning of template.warnings) {
+      console.warn(warning);
+    }
```
@alxhub - is this the best approach to surface these warnings? I couldn't see any other non-fatal mechanism for reporting non-errors.

AndrewKushnir replied:
@petebacondarwin may be add a `TODO(FW-963): ...` to take care of it in a separate PR?

petebacondarwin reviewed (commented):
@AndrewKushnir I worked the PR over quite substantially, so I didn't bother to retain fixup commits. (Sorry). PTAL.

JoostK commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1986,6 +1986,15 @@ export interface ParseTemplateOptions {
+   * expressions.
+   *
+   * If `true` then we will normalize ICU expression line endings.
+   * The default is `false`, but this will be switched in a future major release.
```
Is there a Jira to track this?

petebacondarwin replied:
There is now :-) FW-2106

ayazhafiz commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -204,7 +219,22 @@ class _Tokenizer {
+    const warnings: string[] = [];
+    if (this.nonNormalizedIcuExpressions.length > 0) {
+      warnings.push(
+          `When parsing an external template file "${this._file.url}" ` +
+          'one or more ICU expressions were found that contain `\\r\\n` line-endings.\n' +
+          'These should have been normalized to `\\n` but this would change the computed id for their i18n messages.\n' +
+          'The line endings were left un-normalized for backward compatibility on this occasion.\n' +
+          'In a future release, these line-endings will be normalized, which will be a breaking change.\n' +
+          'It is suggested that you try opting into this behavior now to see what effect it will have on your message ids\n' +
+          'by setting `"normalizeExternalTemplateLineEndingsInICUs": true` in the `angularCompilerOptions` of `tsconfig.json`.\n' +
+          'The ICU expressions that need normalizing are:\n' +
+          this.nonNormalizedIcuExpressions.map(
+              t => ` - "${t.sourceSpan.toString().replace('\n', '\\n').replace('\r', '\\r')}"\n`));
```
bunch of nits
```suggestion
          `When parsing an external template file "${this._file.url}" ` +
          'one or more ICU expressions were found that contain `\\r\\n` line-endings.\n' +
          'These would have been normalized to `\\n` but this would change the computed id for their i18n messages.\n' +
          'The line endings were left unnormalized for backward compatibility on this occasion.\n' +
          'In a future release, these line endings will be normalized, which will be a breaking change.\n' +
          'It is suggested that you try opting into this behavior now to see what effect it will have on your message ids\n' +
          'by setting `"normalizeExternalTemplateLineEndingsInICUs": true` in the `angularCompilerOptions` of `tsconfig.json`.\n' +
          'The ICU expressions that need normalizing are:\n' +
          this.nonNormalizedIcuExpressions.map(
              t => ` - "${t.sourceSpan.toString().replace('\n', '\\n').replace('\r', '\\r')}"\n`));
```

ayazhafiz commented:
LGTM after @JoostK's comments are addressed

petebacondarwin commented on goldens/public-api/compiler-cli/compiler_options.d.ts:
```diff
@@ -7,6 +7,7 @@ export interface I18nOptions {
     enableI18nLegacyMessageIdFormat?: boolean;
     i18nInLocale?: string;
     i18nUseExternalIds?: boolean;
+    normalizeExternalTemplateLineEndingsInICUs?: boolean;
```
Good idea.

petebacondarwin commented:
Marked as `master-only` because it touches the public API (of the compiler options).

AndrewKushnir commented on goldens/public-api/compiler-cli/compiler_options.d.ts:
```diff
@@ -7,6 +7,7 @@ export interface I18nOptions {
     enableI18nLegacyMessageIdFormat?: boolean;
     i18nInLocale?: string;
     i18nUseExternalIds?: boolean;
+    i18nNormalizeExternalTemplateLineEndingsInICUs?: boolean;
```
I'd propose adding a few tests to verify/illustrate that there are no other cases where line endings might become a problem for message extraction and id generation. If we identify other cases, we can consider doing similar changes and reusing this flag (rename it slightly) to provide backwards compatibility.

Some basic cases that I can think of:
- i18n blocks:
```
<div i18n> \r\n
 Some content \r\n
 across multiple lines \r\n
</div>
```
- i18n attributes:
```
<div i18n-title title=" \r\n
Some title \r\n
across multiple lines \r\n
">
```
- i18n ICUs (you have this case in other tests in this PR)
- nested i18n ICUs:
```
{
  messages.length, plural,
  =0 {zero \r\n {type, select, A {type \r\n A} other {OTHER}}}
  =1 {one \r\n {type, select, B {type \r\n B} other {OTHER}}}
}
```
What do you think Pete?

petebacondarwin replied:
Text within blocks have tests already.
I will add for attributes, and for nested ICUs.

petebacondarwin replied:
Or are you suggesting that html_parser_spec tests are not enough and you want to test that message ids are computed correctly still?

AndrewKushnir replied:
Thanks for the reply Pete. My main idea is to check whether we have additional cases like this. I think html_parser_spec that you mentioned gives sufficient level of verification and we can do more checks in a separate task/PR. Thank you.

AndrewKushnir commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -204,7 +219,22 @@ class _Tokenizer {
+          'It is suggested that you try opting into this behavior now to see what effect it will have on your message ids\n' +
+          'by setting `"i18nNormalizeExternalTemplateLineEndingsInICUs": true` in the `angularCompilerOptions` of `tsconfig.json`.\n' +
+          'The ICU expressions that need normalizing are:\n' +
+          this.nonNormalizedIcuExpressions.map(
```
It looks like for nested ICUs we might output the same ICU multiple times, for example:

```
{
  messages.length, plural,
  =0 {zero \r\n {type, select, A {type \r\n A} other {OTHER}}}
  =1 {one \r\n {type, select, B {type \r\n B} other {OTHER}}}
}
```
May be use `Set` instead of array to store problematic ICUs to get rid of duplicates?

petebacondarwin replied:
I'll add a test and try to avoid that.

petebacondarwin replied:
Actually it is preferable to list both instances annotated with their location in the file, since people would need to know where the expressions are that need fixing.

AndrewKushnir reviewed (changes requested):
Thanks for the fix and expanding test coverage @petebacondarwin! 👍

I've added a couple comments to see if we can add extra tests to verify that this problem doesn't exist in other cases that might affect message id generation. Please let me know what you think about it Pete.

Thank you.

IgorMinar commented on goldens/public-api/compiler-cli/compiler_options.d.ts:
```diff
@@ -7,6 +7,7 @@ export interface I18nOptions {
     enableI18nLegacyMessageIdFormat?: boolean;
     i18nInLocale?: string;
     i18nUseExternalIds?: boolean;
+    i18nNormalizeExternalTemplateLineEndingsInICUs?: boolean;
```
my attempt at making this shorter and based on existing naming precedences... what do you think?
```suggestion
    i18nNormalizeExternalTemplateEOLsInICUs?: boolean;
```

I don't love this, because it's still too cryptic and long, also consider `i18nNormalizeEOLsInICUs` - yes, it's more ambiguous but we can clarify in the docs that this applies only to external templates (the definition of "external" will need to be part of the docs anyway, as most people don't understand what "external template" means).

petebacondarwin replied:
My suggestion is `i18nNormalizeLineEndingsInICUs`.
It is not 100% accurate because line endings in ICUs are always normalized for inline templates - and this flag only affects the normalization of ICU expressions in external templates. But I think that it is close enough and short enough.


Igor replied:

> I like your suggestion, but can we also take into consideration that tsconfig.json has a somewhat related precedence for this: https://www.typescriptlang.org/v2/en/tsconfig#newLine we might want to take that into account when deciding the name and intentionally decide to either follow it if the concepts are the same, or intentionally decide to diverge if we are trying to disambiguate between two similar concepts.

petebacondarwin replied:
This is actually not really related that that standard tsconfig setting. That is only about outputting line endings. Whatever the value of that setting, TS always normalizes line endings in backticked strings.

IgorMinar replied:
great thanks for the context in that case `i18nNormalizeLineEndingsInICUs` sgtm

IgorMinar replied:
please re-request the review when the changes are made.

JoostK commented:
Something that came to mind: if there's a lot of components for which a warning is generated, the compilation will generate a lot of long warning messages. Maybe we can find a way to accumulate all occurrences and log them once.

petebacondarwin commented:
> Something that came to mind: if there's a lot of components for which a warning is generated, the compilation will generate a lot of long warning messages. Maybe we can find a way to accumulate all occurrences and log them once.

Actually thinking it, this is a significant issue. There is no straightforward way to aggregate these warnings without making the compiler know more than it ought to about how the warning is structured...

I wonder if we could hide this warning behind a verbose option?
Or perhaps post a simple one line warning and write some documentation for it? Perhaps a link to this PR?
Or perhaps for now we should just not display the warning?

petebacondarwin commented:
I decided to go with not writing any warning to the console at this stage. Apart from anything, this does not really play nice with CLI integration.

JoostK commented on packages/compiler/test/ml_parser/html_parser_spec.ts:
```diff
@@ -309,6 +479,71 @@ import {humanizeDom, humanizeDomSourceSpans, humanizeLineColumn} from './ast_spe
+          expect(parsed.nonNormalizedIcuExpressions).toEqual([]);
+        });
+
+        it('should not normalize line endings in nested expansion forms for external templates',
```
This behaviour is also dependent on the `normalizeExternalTemplateLineEndingsInICUs` flag, no? If so, I would include it in this spec to make it clear that it's still backwards compatible behaviour and not necessarily the desired behaviour.

petebacondarwin replied:
do you mean add `'when `normalizeExternalTemplateLineEndingsInICUs` is not set' to the test description or something else?

JoostK replied:
Yes, that would make it more clear :+1:

JoostK commented on packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:
```diff
@@ -313,6 +313,19 @@ export interface I18nOptions {
+   * If `true` then we will normalize ICU expression line endings.
+   * The default is `false`, but this will be switched in a future major release.
+   */
+  i18nNormalizeLineEndingsInICUs?: boolean;
```
What do you think about adding an ngtsc spec that exercises this flag?

petebacondarwin replied:
Good idea

petebacondarwin replied:
I added integration tests for Ivy i18n that check this.

JoostK reviewed (commented):
Now that no warning is being generated, this should be fine to go into patch once again.

AndrewKushnir commented on packages/compiler/src/ml_parser/parser.ts:
```diff
@@ -23,39 +23,41 @@ export class TreeError extends ParseError {
-  constructor(public rootNodes: html.Node[], public errors: ParseError[]) {}
+  constructor(
+      public rootNodes: html.Node[], public errors: ParseError[],
+      public nonNormalizedIcuExpressions?: lex.Token[]) {}
```
I'm curious if we can leave different forms of parsed ICUs as an internal implementation of Lexer and do not propagate it to Parser (it looks like Parser just proxy them into the final output)?

petebacondarwin replied:
Yes these are just proxying them out. The reason I did this is that I thought that if we were to want to pick up these non-normalized expressions it would be done at the template parser level...
But given that we are not doing anything with them right now, I will remove them from the parser and move the tests for them into the lexer_spec

AndrewKushnir replied:
Thanks Pete, sounds good!

IgorMinar commented on goldens/public-api/compiler-cli/compiler_options.d.ts:
```diff
@@ -6,6 +6,7 @@ export interface BazelAndG3Options {
 export interface I18nOptions {
     enableI18nLegacyMessageIdFormat?: boolean;
     i18nInLocale?: string;
+    i18nNormalizeLineEndingsInICUs?: boolean;
```
I like the name we ended with! Thanks for iterating on it.

IgorMinar replied:
What I was trying to say in https://github.com/angular/angular/pull/36741/files#r415136980 was: could this public API, that is proposed to be an optional prop in this PR, be instead changed to be a prop with a default value? That way we don't need to support the `undefined` value type which is a big win as it makes the API more explicit and self-documenting.

IgorMinar reviewed (commented):
Let's see if we can consider using a default value in the public API instead of undefined.

petebacondarwin commented:
@IgorMinar can you clarify what you mean by:

> Let's see if we can consider using a default value in the public API instead of undefined.

petebacondarwin commented:
> Now that no warning is being generated, this should be fine to go into patch once again.

@JoostK - the `master-only` label is due to the public API change, not because of the warning.

IgorMinar commented:
> @IgorMinar can you clarify what you mean by:
> 
> > Let's see if we can consider using a default value in the public API instead of undefined.

Please take a look at: https://github.com/angular/angular/pull/36741#discussion_r415136980

petebacondarwin committed:
refactor(compiler): tighten type of `TemplateParser._console`

This property can actually be `null` when called from the language-service.
This change allows us to remove the use of `!` to subvert the type system.

petebacondarwin committed:
fix(compiler): normalize line endings in ICU expansions

The html parser already normalizes line endings (converting `\r\n` to `\n`)
for most text in templates but it was missing the expressions of ICU expansions.

In ViewEngine backticked literal strings, used to define inline templates,
were already normalized by the TypeScript parser.
In Ivy we are parsing the raw text of the source file directly so the line
endings need to be manually normalized.

This change ensures that inline templates have the line endings of ICU
expression normalized correctly, which matches the ViewEngine.

In ViewEngine external templates, defined in HTML files, the behavior was
different, since TypeScript was not normalizing the line endings.
Specifically, ICU expansion "expressions" are not being normalized.
This is a problem because it means that i18n message ids can be different on
different machines that are setup with different line ending handling,
or if the developer moves a template from inline to external or vice versa.

The goal is always to normalize line endings, whether inline or external.
But this would be a breaking change since it would change i18n message ids
that have been previously computed. Therefore this commit aligns the ivy
template parsing to have the same "buggy" behavior for external templates.

There is now a compiler option `i18nNormalizeLineEndingsInICUs`, which
if set to `true` will ensure the correct non-buggy behavior. For the time
being this option defaults to `false` to ensure backward compatibility while
allowing opt-in to the desired behavior. This option's default will be
flipped in a future breaking change release.

Further, when this option is set to `false`, any ICU expression tokens,
which have not been normalized, are added to the `ParseResult` from the
`HtmlParser.parse()` method. In the future, this collection of tokens could
be used to diagnose and encourage developers to migrate their i18n message
ids. See FW-2106.

Closes #36725

petebacondarwin committed:
refactor(compiler): simplify tokenizer and parser results

Move the creation of the results objects into the wrapper functions.
This makes it easier to reason about what the parser and lexer classes
are responsible for - you create a new object for each tokenization or
parsing activity and they hold the state of the activity.

petebacondarwin committed:
test(compiler): fix unwanted `!` operators

The `I18nComponent` was using `!` for some of its properties
because it had not initialized them. This is now resolved by explictly
marking them as optional.

petebacondarwin committed:
test(compiler): check extracted translations for line endings

When there are `CRLF` line endings some translations get different
message ids. This commit adds tests to track this in ViewEngine.

petebacondarwin commented:
I added integration tests for ViewEngine and Ivy that check the generated message ids.

IgorMinar commented on packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -96,8 +96,9 @@ export class DecorationAnalyzer {
         /* i18nUseExternalIds */ true, this.bundle.enableI18nLegacyMessageIdFormat,
-        this.moduleResolver, this.cycleAnalyzer, this.refEmitter, NOOP_DEFAULT_IMPORT_RECORDER,
-        NOOP_DEPENDENCY_TRACKER, this.injectableRegistry, /* annotateForClosureCompiler */ false),
+        /* normalizeExternalTemplateLineEndingsInICUs */ false, this.moduleResolver,
```
plz update the comment with the new prop name

petebacondarwin replied:
This is actually an internal argument name for the `ComponentHandler`. And I thought we might leave it as the more explicit name, since it is not public. But given that you mentioned it, I am happy to change it to be in sync with the public property.

IgorMinar reviewed (approved):
LGTM! I have no more concerns except for that one inline comment update. Thanks!

Reviewed-for: public-api

gkalpak reviewed (approved):
Reviewed-for: fw-ngcc

AndrewKushnir commented:
[Presubmit](http://test/OCL:308852114:BASE:308852155:1588095793261:2278bccc)
